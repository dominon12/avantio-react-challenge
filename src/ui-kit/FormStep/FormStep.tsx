import type { FormEventHandler, ReactNode, Ref } from "react";

import Title from "../Title";
import Button from "../Button";

type FormStepProps = {
  // new way to pass refs since React 19
  ref?: Ref<HTMLFormElement>;
  title: string;
  children: ReactNode;
  submitText?: string;
  canSubmit?: boolean;
  onSubmit: FormEventHandler<HTMLFormElement>;
  onBack?: () => void;
};

/**
 * Renders a form.
 */
function FormStep(props: FormStepProps) {
  const {
    ref,
    title,
    children,
    submitText = "Submit",
    canSubmit = true,
    onSubmit,
    onBack,
  } = props;

  return (
    <section className="flex h-full flex-col">
      <Title>{title}</Title>
      <form ref={ref} onSubmit={onSubmit} className="flex grow flex-col">
        <div className="mb-8 flex grow flex-col gap-3">{children}</div>

        <div className="flex gap-2">
          {onBack && (
            <Button className="grow" type="button" onClick={onBack}>
              Back
            </Button>
          )}
          <Button className="grow" type="submit" disabled={!canSubmit}>
            {submitText}
          </Button>
        </div>
      </form>
    </section>
  );
}

export default FormStep;
