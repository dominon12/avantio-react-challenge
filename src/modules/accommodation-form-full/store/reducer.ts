import FormStep from "../types/form-step";
import Action from "./action";
import ActionType from "./action-type";
import initialState from "./initial-state";
import StoreState from "./state-type";

function reducer(state: StoreState, action: Action): StoreState {
  switch (action.type) {
    case ActionType.AccommodationFormSubmitted:
      return {
        ...state,
        step: FormStep.Owner,
        accommodation: action.payload,
      };
    case ActionType.OwnerFormSubmitted:
      return {
        ...state,
        step: FormStep.Summary,
        owner: action.payload,
      };
    case ActionType.FormSummaryAccepted:
      return {
        ...state,
        step: FormStep.Confirmation,
        result: action.payload.result,
      };
    case ActionType.ResetForm:
      return initialState;
    default:
      return state;
  }
}

export default reducer;
