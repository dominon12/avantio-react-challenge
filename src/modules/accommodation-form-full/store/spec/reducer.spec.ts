import { describe, expect, it } from "vitest";

import type Action from "../action";
import ActionType from "../action-type";
import mockAccommodation from "src/tests/mock-data/mock-accommodation";
import reducer from "../reducer";
import initialState from "../initial-state";
import FormStep from "../../types/form-step";
import mockOwner from "src/tests/mock-data/mock-owner";
import SubmissionResult from "src/modules/submission-confirmation/types/submission-result";

describe("Accommodation form reducer", () => {
  it("stores accommodation and goes to the `Owner` step", () => {
    // prepare action
    const action: Action = {
      type: ActionType.AccommodationFormSubmitted,
      payload: mockAccommodation,
    };

    // call reducer
    const result = reducer(initialState, action);

    // check result
    expect(result.step).toBe(FormStep.Owner);
    expect(result.accommodation).toEqual(mockAccommodation);

    // expect other state properties to be unchanged
    expect(result.owner).toBeUndefined();
    expect(result.result).toBeUndefined();

    // match snapshot
    expect(result).toMatchSnapshot();
  });

  it("stores owner and goes to the `Summary` step", () => {
    // prepare action
    const action: Action = {
      type: ActionType.OwnerFormSubmitted,
      payload: mockOwner,
    };

    // call reducer
    const result = reducer(initialState, action);

    // check result
    expect(result.step).toBe(FormStep.Summary);
    expect(result.owner).toEqual(mockOwner);

    // expect other state properties to be unchanged
    expect(result.accommodation).toBeUndefined();
    expect(result.result).toBeUndefined();

    // match snapshot
    expect(result).toMatchSnapshot();
  });

  it("stores result and goes to the `Confirmation` step", () => {
    const mockResult = SubmissionResult.Success;

    // prepare action
    const action: Action = {
      type: ActionType.FormSummaryAccepted,
      payload: { result: mockResult },
    };

    // call reducer
    const result = reducer(initialState, action);

    // check result
    expect(result.step).toBe(FormStep.Confirmation);
    expect(result.result).toEqual(mockResult);

    // expect other state properties to be unchanged
    expect(result.accommodation).toBeUndefined();
    expect(result.owner).toBeUndefined();

    // match snapshot
    expect(result).toMatchSnapshot();
  });

  it("returns the initial state for an unknown action type", () => {
    // prepare action
    const action = {
      type: "UNKNOWN_ACTION",
    } as unknown as Action;

    // call reducer
    const result = reducer(initialState, action);

    // check result
    expect(result).toBe(initialState);

    // match snapshot
    expect(result).toMatchSnapshot();
  });
});
