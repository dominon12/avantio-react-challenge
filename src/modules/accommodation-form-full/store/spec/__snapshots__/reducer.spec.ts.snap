// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Accommodation form reducer > resets the state correctly 1`] = `
{
  "step": "accommodation",
}
`;

exports[`Accommodation form reducer > returns the initial state for an unknown action type 1`] = `
{
  "step": "accommodation",
}
`;

exports[`Accommodation form reducer > stores accommodation and goes to the \`Owner\` step 1`] = `
{
  "accommodation": {
    "address": "123 Test St",
    "name": "Test Accommodation",
    "type": "house",
  },
  "step": "owner",
}
`;

exports[`Accommodation form reducer > stores owner and goes to the \`Summary\` step 1`] = `
{
  "owner": {
    "email": "test@email.com",
    "name": "Test Owner",
  },
  "step": "summary",
}
`;

exports[`Accommodation form reducer > stores result and goes to the \`Confirmation\` step 1`] = `
{
  "result": "success",
  "step": "confirmation",
}
`;
